dipy/tracking/streamline.py

```diff
@@ -323,3 +327,134 @@ def orient_by_rois(streamlines, roi1, roi2, affine=None, copy=True):
+        volume.
+
+    streamlines : ndarray or list
+        If array, of shape (n_streamlines, n_nodes, 3)
+        If list, len(n_streamlines) with (n_nodes, 3) array in
+        each element of the list.
+
+    affine : ndarray, shape (4, 4)
+        Affine transformation from voxels (image coordinates) to streamlines.
+        Default: identity.
+
+    threedvec : bool
+        Whether the last dimension has length 3. This is a special case in
+        which we can use :func:`vfu.interpolate_vector_3d` for the
+        interploation of 4D volumes without looping over the elements of the
+        last dimension.
+
+    Return
+    ------
+    array or list (depending on the input) : values interpolate to each
+        coordinate along the length of each streamline
+    """
+    data = data.astype(np.float)
+    if affine is not None:
+        streamlines = ut.move_streamlines(streamlines,
+                                          np.linalg.inv(affine))
+
+    if (isinstance(streamlines, list) or
+            isinstance(streamlines, types.GeneratorType)):
+        vals = []
+        for sl in streamlines:
+
+            if threedvec:
+                vals.append(list(vfu.interpolate_vector_3d(data,
+                                 sl.astype(np.float))[0]))
+            else:
+                vals.append(list(vfu.interpolate_scalar_3d(data,
+                                 sl.astype(np.float))[0]))
+
+    elif isinstance(streamlines, np.ndarray):
+        sl_shape = streamlines.shape
+        sl_cat = streamlines.reshape(sl_shape[0] *
+                                     sl_shape[1], 3).astype(np.float)
+        if affine is not None:
+            sl_cat = (np.dot(sl_cat, affine[:3, :3]) + affine[:3, 3])
+
+        # So that we can index in one operation:
+        if threedvec:
+            vals = np.array(vfu.interpolate_vector_3d(data, sl_cat)[0])
+        else:
+            vals = np.array(vfu.interpolate_scalar_3d(data, sl_cat)[0])
+        vals = np.reshape(vals, (sl_shape[0], sl_shape[1], -1)).squeeze()
+
+    else:
+        raise RuntimeError("Extracting values from a volume ",
+                           "requires streamlines input as an array, ",
+                           "a list of arrays, or a streamline generator.")
+
+    return vals
+
+
+def values_from_volume(data, streamlines, affine=None):
+    """Extract values of a scalar/vector along each streamline from a volume.
+
+    Parameters
+    ----------
+    data : 3D or 4D array
+        Scalar (for 3D) and vector (for 4D) values to be extracted. For 4D
+        data, interpolation will be done on the 3 spatial dimensions in each
+        volume.
+
+    streamlines : ndarray or list
+        If array, of shape (n_streamlines, n_nodes, 3)
+        If list, len(n_streamlines) with (n_nodes, 3) array in
+        each element of the list.
+
+    affine : ndarray, shape (4, 4)
+        Affine transformation from voxels (image coordinates) to streamlines.
```

Comment:
So its the streamlines which are moved (according to the function) to the
voxel space of the metrics. I was just wondering if the term used are
enough to guess what is happening, pretty new to all these confusing
conventions, but if people who know what they are doing understands what
you mean, then I'm good (and will read more stuff)

2016-01-25 17:35 GMT+01:00 Ariel Rokem <notifications@github.com>:

> In dipy/tracking/streamline.py
> <https://github.com/nipy/dipy/pull/811#discussion_r50718149>:
>
> > +    """Extract values of a scalar/vector along each streamline from a volume.
> > +
> > +    Parameters
> > +    ----------
> > +    data : 3D or 4D array
> > +        Scalar (for 3D) and vector (for 4D) values to be extracted. For 4D
> > +        data, interpolation will be done on the 3 spatial dimensions in each
> > +        volume.
> > +
> > +    streamlines : ndarray or list
> > +        If array, of shape (n_streamlines, n_nodes, 3)
> > +        If list, len(n_streamlines) with (n_nodes, 3) array in
> > +        each element of the list.
> > +
> > +    affine : ndarray, shape (4, 4)
> > +        Affine transformation from voxels (image coordinates) to streamlines.
>
> It's whatever you did. If you represented your streamlines in "world
> coordinates" (mm, relative to scanner iso-center), then you need an affine
> to transform from that to the image.
> … <#370173273_>
> On Mon, Jan 25, 2016 at 8:18 AM, Samuel St-Jean <notifications@github.com>
> wrote: In dipy/tracking/streamline.py <#811 (comment)
> <https://github.com/nipy/dipy/pull/811#discussion_r50715670>>: > +
> """Extract values of a scalar/vector along each streamline from a volume. >
> + > + Parameters > + ---------- > + data : 3D or 4D array > + Scalar (for
> 3D) and vector (for 4D) values to be extracted. For 4D > + data,
> interpolation will be done on the 3 spatial dimensions in each > + volume.
> > + > + streamlines : ndarray or list > + If array, of shape
> (n_streamlines, n_nodes, 3) > + If list, len(n_streamlines) with (n_nodes,
> 3) array in > + each element of the list. > + > + affine : ndarray, shape
> (4, 4) > + Affine transformation from voxels (image coordinates) to
> streamlines. streamlines space is world cooordinate or some other space? —
> Reply to this email directly or view it on GitHub <
> https://github.com/nipy/dipy/pull/811/files#r50715670>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/nipy/dipy/pull/811/files#r50718149>.
>